Add tropical linear spaces from graphs

YueRen opened the pull request:
This pull requests adds a new constructor for tropical linear spaces from graphs.  Useful (for me) for example for graph rigidity questions (see e.g. [here](https://arxiv.org/abs/2201.00533)).

YueRen committed:
TropicalGeometry: tropical_linear_space from graphs

YueRen commented:
@ollieclarke8787 Would you be able to review this pull request?

lgoettgens commented on src/TropicalGeometry/linear_space.jl:
```diff
@@ -365,6 +365,34 @@ function tropical_linear_space(I::MPolyIdeal, nu::Union{Nothing,TropicalSemiring
+    for (i,edge) in enumerate(edges(G))
+        M[src(edge),i] = 1
+        M[dst(edge),i] = -1
+    end
```
We already have `signed_incidence_matrix`. Could you maybe use that here?

YueRen replied:
`signed_incidence_matrix` only works for directed graphs.  What I have is an undirected graph and I need the signed incidence matrix for a choice of directions (the choice does not matter).

lgoettgens replied:
That's still something worth putting in its own function for people to use. I don't know if polymake has an acessor for that as well or if this would need your simple loop. Maybe @benlorenz or @lkastner can help out?

YueRen replied:
True, this is actually one of the changes of the Durham workshop that I am preparing right now.

YueRen replied:
Once https://github.com/oscar-system/Oscar.jl/pull/4450 is merged, I am happy to update it.

YueRen replied:
Okay, is fixed in the newest version.

ollieclarke8787 reviewed (approved):
Code correctly produces Bergman fan of the cycle matroid the graph. The example of K_3 is also correct.

YueRen commented:
@benlorenz I've changed `scalar_types_extended`:

```
-const scalar_types_extended = Union{scalar_types,ZZRingElem}
+const scalar_types_extended = Union{scalar_types,IntegerUnion}
```

However, it creates a bunch of errors in the type tests that I don't know how to fix:
https://github.com/oscar-system/Oscar.jl/actions/runs/12946690999/job/36111735653?pr=4444#step:8:746

Can you please look into it when you have the time (most likely not during the currently ongoing polymake developer meeting)?  Many thanks in advance!

fingolfin commented:
I am confused, as this PR does not touch `scalar_types_extended` ?!?

Anyway, the error in the linked log suggest that more things need to be changed if you adjust `scalar_types_extended`. Presumably if an `Int` is now used in places, them at some point some code may call `parent` on it, and this yields `Integers{Int64}()` -- a special placeholder parent for Julia integers that AA provides. And apparently this is then passed as first argument to `point_vector` which doesn't know about it.

thofma commented:
I think Yue refers to #4444

benlorenz commented on src/TropicalGeometry/linear_space.jl:
```diff
@@ -365,6 +365,30 @@ function tropical_linear_space(I::MPolyIdeal, nu::Union{Nothing,TropicalSemiring
+@doc raw"""
+    tropical_linear_space(G::Graph, nu::TropicalSemiringMap; weighted_polyhedral_complex_only::Bool=false)
+
+Return the Bergman fan of the graphic matroid of `G` as a tropical linear space, the tropical semiring map `nu` is used to fix the convention.  If `weighted_polyhedral_complex==true`, will not cache any extra information.
```
```suggestion
Return the Bergman fan of the graphic matroid of `G` as a tropical linear space, the tropical semiring map `nu` is used to fix the convention.  If `weighted_polyhedral_complex_only` is true, will not cache any extra information.
```

benlorenz commented on src/TropicalGeometry/linear_space.jl:
```diff
@@ -365,6 +365,30 @@ function tropical_linear_space(I::MPolyIdeal, nu::Union{Nothing,TropicalSemiring
 
 
+@doc raw"""
+    tropical_linear_space(G::Graph, nu::TropicalSemiringMap; weighted_polyhedral_complex_only::Bool=false)
```
```suggestion
    tropical_linear_space(G::Graph[, nu::TropicalSemiringMap]; weighted_polyhedral_complex_only::Bool=false)
```
Maybe mention that this is optional (and what it defaults to)?

micjoswig replied:
I think that the design is currently not really consistent.

The [doc intro](https://github.com/oscar-system/Oscar.jl/blob/master/docs/src/TropicalGeometry/intro.md) says "Like [[MS15](https://docs.oscar-system.org/dev/references/#MS15)], OSCAR follows the min-convention by default and unless specified otherwise."  Yet the code, e.g., [here](https://github.com/oscar-system/Oscar.jl/blob/d9efcbce878eb3bad9c3b220edd42b723c0add75/src/TropicalGeometry/initial.jl#L25) does not set any default.

The new code on tropical linear spaces, however, seems to force min as the default.  So it is consistent with the docs but not with the rest of the code.

YueRen committed:
Update src/TropicalGeometry/linear_space.jl

Co-authored-by: Benjamin Lorenz <benlorenz@users.noreply.github.com>

fingolfin commented:
I am not sure who from the inside is able to review (and approve!) this tropical geometry PR. Perhaps someone from Berlin, e.g. @lkastner @benlorenz @micjoswig ?

micjoswig commented on docs/src/TropicalGeometry/linear_space.md:
The docs on tropical linear spaces (and the rest of tropical geometry) need to state explicitly where the various objects (tropical varieties, tropical linear spaces etc) actually live.  Candidates include R^n or the tropical projective torus R^n/R1 (in the homogeneous setting).  Either there are choices for the user to make, or there are fixed global decisions.

Specifically, tropical linear spaces may also be viewed as tropical polytopes.  And then there are more choices, for they can be considered in a tropical projective torus or in a tropical projective space.  I guess, none of this is currently supported in OSCAR.

The role of Bergman fans and constant vs non-constant coefficients must be explained.

YueRen replied:
Good points, I will adjust the documentation accordingly.

YueRen committed:
TropicalGeometry: changed docstring to indicate optional TropicalSemiringMap

YueRen committed:
TropicalGeometry: added default tropical_semiring_map for consistency

This is technically not necessary, as passing `nothing` to
tropical_linear_space inside the function is valid, but this makes it
more consistent and maybe easier to read

YueRen committed:
TropicalGeometry: changed linear space docu

Added a note that tropical linear spaces in OSCAR (currently) live in
euclidean space and not in the tropical torus

YueRen committed:
TropicalGeometry: streamlined default TropicalSemiringMap declarations

YueRen committed:
TropicalGeometry: improved tropical_matrix

YueRen committed:
TropicalGeometry: updated tropical_linear_space test

Constructing tropical linear spaces over the rationals go through an
optimised polymake function and thus they lack certain
attributes (because computing them would be extra work)

YueRen committed:
TropicalGeometry: fixed variable naming

YueRen committed:
TropicalGeometry: compute algebraic matrix from algebraic ideal if not cached

YueRen committed:
TropicalGeometry: fixed typo in docs

YueRen committed:
TropicalGeometry: added note on Bergman fans

YueRen commented:
Besides enabling the constructing of tropical linear spaces from graphs, the following have changed:

1. Replaced how optional `TropicalSemiringMap` are written.  Before the function signature included `nu::Union{Nothing,TropicalSemiringMap}=nothing` and the function body included code that set `nu` when `nu==nothing`.  Now this is done in the function signature directly like:

```
function tropical_linear_space(I::MPolyIdeal, nu::TropicalSemiringMap=tropical_semiring_map(coefficient_ring(I)); weighted_polyhedral_complex_only::Bool=false)
```

2. Adding a few sentences to the documentation, that emphasize that 
- tropical linear spaces in OSCAR live in the euclidean space, not the tropical torus
- tropical linear spaces under a trivial valuation are Bergman fans and they may have different structures depending on how they are constructed

micjoswig reviewed (approved):
I think it is OK now.

Some minor edits, e.g., improving the docs further, can be made after the merge.  I find that easier than messing with the code in an external repo.
